localstack/services/s3/provider_stream.py

```diff
@@ -242,14 +242,16 @@ def copy_object(
         )
         if checksum_algorithm:
             # this is a bug in AWS: it sets the content encoding header to an empty string (parity tested)
-            dest_key_object = get_key_from_moto_bucket(dest_moto_bucket, key=request["Key"])
+            dest_key_object: StreamedFakeKey = get_key_from_moto_bucket(
+                dest_moto_bucket, key=request["Key"]
+            )
             dest_key_object.checksum_algorithm = checksum_algorithm
 
             if not source_key_object.checksum_value:
                 stream_value: SpooledTemporaryFile = dest_key_object.value
                 checksum = get_s3_checksum(checksum_algorithm)
 
-                with dest_key_object.lock:
+                with dest_key_object.lock, dest_key_object.read_lock:
```

Comment:
Q: i think i may be missing something fundamental. why do we need both locks here and in all the other blocks?